exporter/awsemfexporter/cwlog_client.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsemfexporter
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs/cloudwatchlogsiface"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter/handler"
+)
+
+const (
+	// this is the retry count, the total attempts would be retry count + 1 at most.
+	defaultRetryCount          = 5
+	ErrCodeThrottlingException = "ThrottlingException"
+)
+
+var (
+	// backoff retry 6 times
+	sleeps = []time.Duration{
+		time.Millisecond * 200, time.Millisecond * 400, time.Millisecond * 800,
+		time.Millisecond * 1600, time.Millisecond * 3200, time.Millisecond * 6400}
+)
+
+//The log client will perform the necessary operations for publishing log events use case.
+type LogClient interface {
+	PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) *string
+	CreateStream(logGroup, streamName *string) (token string, e error)
+}
+
+// Possible exceptions are combination of common errors (https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/CommonErrors.html)
+// and API specific erros (e.g. https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html#API_PutLogEvents_Errors)
+type CloudWatchLogClient struct {
+	svc    cloudwatchlogsiface.CloudWatchLogsAPI
+	logger *zap.Logger
+}
+
+//Create a log client based on the actual cloudwatch logs client.
+func newCloudWatchLogClient(svc cloudwatchlogsiface.CloudWatchLogsAPI, logger *zap.Logger) *CloudWatchLogClient {
+	logClient := &CloudWatchLogClient{svc: svc,
+		logger: logger}
+	return logClient
+}
+
+// NewCloudWatchLogsClient create CloudWatchLogClient
+func NewCloudWatchLogsClient(logger *zap.Logger, awsConfig *aws.Config, sess *session.Session) LogClient {
+	client := cloudwatchlogs.New(sess, awsConfig)
+	client.Handlers.Build.PushBackNamed(handler.RequestStructuredLogHandler)
+	return newCloudWatchLogClient(client, logger)
+}
+
+//Put log events. The method mainly handles different possible error could be returned from server side, and retries them
+//if necessary.
+func (client *CloudWatchLogClient) PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) *string {
```

Comment:
Return error to collector pipeline instead of handling everything here. Requests that should not be retried (HTTP 4xx) must be wrapped in consumererror.Permanent